pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go

```diff
@@ -314,44 +342,76 @@ func (c *DisruptionBudgetController) execute(key string) error {
 	}
 
 	// Only consider pdbs which belong to this vmi
-	pdb, err := c.pdbForVMI(namespace, name)
+	pdbs, err := c.pdbsForVMI(namespace, name)
 	if err != nil {
 		log.DefaultLogger().Reason(err).Error("Failed to fetch pod disruption budgets for namespace from cache.")
 		// If the situation does not change there is no benefit in retrying
 		return nil
 	}
 
-	return c.sync(key, vmi, pdb)
+	if len(pdbs) == 0 {
+		return c.sync(key, vmi, nil)
+	}
 
+	for i := range pdbs {
+		if syncErr := c.sync(key, vmi, pdbs[i]); syncErr != nil {
+			err = syncErr
+		}
+	}
+	return err
 }
 
-func (c *DisruptionBudgetController) sync(key string, vmi *virtv1.VirtualMachineInstance, pdb *v1beta1.PodDisruptionBudget) error {
-	delete := false
-	create := false
-	// delete if there is no VMI
-	if vmi == nil && pdb != nil {
-		delete = true
-	} else if vmi != nil {
-		wantsToMigrate := wantsToMigrateOnDrain(vmi)
-		if vmi.DeletionTimestamp != nil && pdb != nil {
-			// pdb can already be deleted, shutdown already in process
-			delete = true
-		} else if !wantsToMigrate && pdb != nil {
-			// We don't want migrations on evictions, if there is a pdb, remove it
-			delete = true
-		} else if wantsToMigrate && vmi.DeletionTimestamp == nil && pdb == nil {
-			// No pdb and we want migrations on evictions
-			create = true
-		} else if wantsToMigrate && pdb != nil {
-			if ownerRef := v1.GetControllerOf(pdb); ownerRef != nil && ownerRef.UID != vmi.UID {
-				// The pdb is from an old vmi with a different uid, delete and later create the correct one
-				// The VMI always has a minimum grace period, so normally this should not happen, therefore no optimizations
-				delete = true
-			}
+func (c *DisruptionBudgetController) isMigrationComplete(vmi *virtv1.VirtualMachineInstance, migrationName string) (bool, error) {
+	objs, err := c.migrationInformer.GetIndexer().ByIndex(cache.NamespaceIndex, vmi.Namespace)
+	if err != nil {
+		return false, err
+	}
+
+	var migration *virtv1.VirtualMachineInstanceMigration
+	for _, obj := range objs {
+		vmim := obj.(*virtv1.VirtualMachineInstanceMigration)
+		if vmim.GetName() == migrationName {
+			migration = vmim
+			break
+		}
+	}
+
+	if migration == nil {
+		// if no migration is found we consider it as completed
+		return true, nil
+	} else if !(migration.IsFinal() || (vmi.Status.MigrationState != nil && vmi.Status.MigrationState.MigrationUID == migration.UID && vmi.Status.MigrationState.Completed)) {
+		return false, nil
+	}
+
+	objs, err = c.podInformer.GetIndexer().ByIndex(cache.NamespaceIndex, vmi.Namespace)
+	if err != nil {
+		return false, err
+	}
+
+	runningPods := 0
+	for _, obj := range objs {
+		pod := obj.(*corev1.Pod)
+		vmiUid := pod.Labels[virtv1.CreatedByLabel]
+
+		if vmiUid == string(vmi.UID) && pod.Status.Phase == corev1.PodRunning {
+			runningPods += 1
```

Comment:
take a look at the `pkg/controller/controller.go` files VMIActivePodsCount() function. I think it gives you exactly this.

Reply:
great didn't notice that existed :)